Restructuring and adding testing

rjzamora opened the pull request:
Restructuring, in response to issue#5 (https://github.com/gpuopenanalytics/pynvml/issues/5).  I also added some simple testing and changed the names of several classes/methods in `pynvml/nvml.py` to be more "pythonic".  Most of the nvml wrapper functions still have the same name as the C library, so more work is needed if we want a conventional python interface.

rjzamora committed:
starting restructuring and cleanup

rjzamora committed:
more restructuring and testing

rjzamora committed:
starting pynvml name cleaup -- cleanup needed in nvidia_smi

rjzamora committed:
adding tests for smi wrapper

mrocklin commented on pynvml/tests/test_smi.py:
```diff
@@ -0,0 +1,77 @@
+# Test memory-utilization query
+def test_memory_utilization(get_count):
+    def _memory_util_query(j):
+        return pytest.nvsmi.DeviceQuery('utilization.memory')['gpu'][j]['utilization']['memory_util']
```
I'm curious, why separate this line out into a separate function definition?  Why not use it directly in the for loop below?

rjzamora replied:
Right - I separated the query into a function only because the query syntax was a bit verbose for some cases.  Beyond making things easier for me to read, there is no good reason to do this.

mrocklin replied:
OK, I've submitted https://github.com/rjzamora/pynvml/pull/1 as a proposed alternate.  

I change both how we handle fixtures (which I think we should change to match current idioms) and the pattern used here for tests (which is subjective style, but I thought I'd put it out anyway).

mrocklin replied:
We'll want to change the existing tests here to the new style I think.

mrocklin commented on pynvml/tests/test_smi.py:
```diff
@@ -0,0 +1,77 @@
+        return pytest.nvsmi.DeviceQuery()['gpu'][j]['pci']['rx_util']
+    for i in range(pytest.ngpus):
+        rx_util = _rx_util_query(i)
+        assert(rx_util >= 0)
```
These two tests take the most time, I suspect because they grab all of the information and then filter it down, rather than just requesting the rx_util/tx_util information.  Is there a way to slim these down?

```
(ucx-dev) mrocklin@dgx16:~/pynvml$ py.test pynvml/tests/ -s -v --durations 10
=============================================================== test session starts ===============================================================
platform linux -- Python 3.7.3, pytest-4.4.1, py-1.8.0, pluggy-0.9.0 -- /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python
cachedir: .pytest_cache
rootdir: /home/nfs/mrocklin/pynvml
plugins: asyncio-0.10.0
collected 17 items

pynvml/tests/test_nvml.py::test_nvmlSystemGetNVMLVersion [NVML Version: 396.44] PASSED
pynvml/tests/test_nvml.py::test_nvmlSystemGetProcessName [Process: /home/nfs/mrocklin/miniconda/envs/ucx-dev/bin/python] PASSED
pynvml/tests/test_nvml.py::test_nvmlSystemGetDriverVersion [Driver Version: 396.44] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetHandleBySerial [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetHandleByUUID [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetHandleByPciBusId [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetPowerUsage [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetMemoryInfo [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetUtilizationRates [8 GPUs] PASSED
pynvml/tests/test_nvml.py::test_nvmlDeviceGetPcieThroughput [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_query_memory [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_gpu_utilization [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_memory_utilization [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_tx_util [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_rx_util [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_pstate [8 GPUs] PASSED
pynvml/tests/test_smi.py::test_temperature [8 GPUs] PASSED

============================================================ slowest 10 test durations ============================================================
12.61s call     pynvml/tests/test_smi.py::test_rx_util
12.18s call     pynvml/tests/test_smi.py::test_tx_util
0.49s call     pynvml/tests/test_nvml.py::test_nvmlDeviceGetPcieThroughput
0.26s call     pynvml/tests/test_smi.py::test_temperature
0.06s setup    pynvml/tests/test_smi.py::test_query_memory
0.05s setup    pynvml/tests/test_nvml.py::test_nvmlDeviceGetHandleBySerial
0.03s call     pynvml/tests/test_smi.py::test_query_memory
0.02s call     pynvml/tests/test_smi.py::test_gpu_utilization
0.02s setup    pynvml/tests/test_nvml.py::test_nvmlSystemGetNVMLVersion
0.02s call     pynvml/tests/test_nvml.py::test_nvmlDeviceGetPowerUsage
```

mrocklin reviewed (commented):
In general this looks good to me.  I added a couple of comments/questions inline in the tests.

I can imagine that we might also want to get a bit more opinionated in the future about how to start things off.  The current approach here is probably overly nested for a naive user to discover correct behavior without guidance.

```
from pynvml import smi
s = smi.nvidia_smi.getInstance()
s.DeviceQuery(...)
```

I'll raise another issue about that in a bit though.

Also, for standard code styling, you might want to check out [black](https://github.com/ambv/black).  I've been starting to use it in projects and quite enjoy it.

mrocklin committed:
Restructure nvidia-smi fixture usage and example test

This uses return values from the fixtures, rather than attaching things to the
Pytest module, which is a bit more idiomatic.  It also prefers shorter names.

Then we change around a single test to serve as an example.
This does a single call and avoids the indirection of a function definition.

rjzamora committed:
Merge pull request #1 from mrocklin/restructure-fixtures

Restructure nvidia-smi fixture usage and example test

mrocklin commented on README.md:
```diff
@@ -25,25 +25,25 @@ Usage
->>> from pynvml import *
->>> nvmlInit()
+>>> from pynvml.nvml import *
+>>> nvml_init()
```
Did these actually change in our version of the code?  Or were these docstrings just very out of date?

rjzamora replied:
Yeah - We did change the structure and the name of the initialization and finalization function wrappers.  Since we left all the other function-wrapper names unchanged, it might make sense to leave `nvmlInit()` and `nvmlShutdown()` alone for now (but all the names should probably be changed eventually)

mrocklin commented on README.md:
````diff
@@ -25,25 +25,25 @@ Usage
 ```python
->>> from pynvml import *
->>> nvmlInit()
+>>> from pynvml.nvml import *
````
It's a bit ugly, but you probably want to add 

```python
from .nvml import *
```

In `pynvml/__init__.py`

Then you can drop the `.nvml` 

(also, just as an FYI, using `import *` is considered a bit odd in Python these days, though I'm not sure how better to handle a library with such a large namespace like this.  Probably one should just `import pynvml` and use `pynvml.foo` everywhere, but that's a topic for another PR I think).

mrocklin commented on pynvml/tests/test_nvml.py:
```diff
@@ -0,0 +1,210 @@
+
+# Get GPU count
+@pytest.fixture
+def get_gpu_count(nvml_open):
```
It would be good to also change these fixtures as well so that they return values rather than modify the `pytest` module, as was done in the other testing file.

mrocklin reviewed (commented):
OK, I think that there are a few (hopefully staightforward) changes to make.  Then we can merge this and cut a release.  I suspect that this may be useful for some of the device-host-disk spill work.

rjzamora committed:
Addressing most code-review concerns.  Reverting any changes that alter how a user must interact with pynvml or nvidia_smi (adding back nvmlInit, nvmlShutdown, etc). There is further restructuring and test cleanup.

rjzamora commented:
@mrocklin Okay - I think I addressed the testing.  Also, the pynvml and nvida_smi modules can now be loaded and used exactly as before.  We will probably want to improve the way the user interacts with these modules in the future, but I think we should let this PR focus on restructuring and testing.

mrocklin commented:
Thanks @rjzamora .  This is in!
